### PR TITLE
Feat name constraints

### PR DESCRIPTION
**Pull request by gfxcc**
Add support for X.509 name constraints.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
privateca: added support for X.509 name constraints
```

### CODE REVIEW

File: mmv1/products/privateca/api.yaml [hunk 1]
```diff
@@ -328,6 +328,88 @@ objects:
                           item_type: Api::Type::Integer
                           description: |
                             An ObjectId specifies an object identifier (OID). These provide context and describe types in ASN.1 messages.
+              - !ruby/object:Api::Type::NestedObject
+                name: 'nameConstraints'
```
Comment (shuyama1) [inline mmv1/products/privateca/api.yaml:332]: Do you think if it makes sense to mark one of the subfields required, likely `critical`? This way, we can prevent user from sending empty block `name_constraints{}`, as it usually end up causing more problems.

**Review by shuyama1 — COMMENTED** [review]
LGTM overall! Only one small comment.

We recently split the product level api.yaml to individual resource files. You'll likely need to move the code in `api.yaml` to each resource `.yaml` file in https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/products/privateca to resolve conflicts. Sorry for the extra work.

File: mmv1/products/privateca/api.yaml (hunk 1 above)
Comment (gfxcc) [reply mmv1/products/privateca/api.yaml]: Can customer set a `required` boolean field to `false`?

Comment (shuyama1) [reply mmv1/products/privateca/api.yaml]: yeah, I believe they should be able to

Comment (gfxcc) [reply mmv1/products/privateca/api.yaml]: Done.

**Comment by gfxcc** [comment]
> LGTM overall! Only one small comment.
> 
> We recently split the product level api.yaml to individual resource files. You'll likely need to move the code in `api.yaml` to each resource `.yaml` file in https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/products/privateca to resolve conflicts. Sorry for the extra work.

Done. The resource split looks great! That does make the code more easier to be managed.

PTAL

File: mmv1/products/privateca/CaPool.yaml [hunk 2]
```diff
@@ -343,6 +343,79 @@ properties:
...
+                The value can be a hostname or a domain with a
+                leading period (like `.example.com`)
+              item_type: Api::Type::String
+            - !ruby/object:Api::Type::Array
+              name: 'excludedUris'
+              description: |
+                Contains the excluded URIs that apply to the host part of the name.
+                The value can be a hostname or a domain with a
+                leading period (like `.example.com`)
+              item_type: Api::Type::String
```
[the remark this answers is not in the document]
Comment (gfxcc) [reply mmv1/products/privateca/CaPool.yaml:418]: > * Can we test multiple entries for these values? Just want to make sure if the API returns the elements the same order they've been configured to make sure if they're list or set.

Done. 

> * Do we know if API will return values for these fields if they're not explicitly specified? My assumption is no, but just want to make sure.

No.

**Review by shuyama1 — APPROVED** [review]
Failing test is not related to this PR. Thank you for your contribution!

**Pull request merged**